apps/gdalalg_vector_geom_simplify_coverage.cpp

```diff
@@ -0,0 +1,316 @@
+    AddArg("tolerance", 0, _("Distance tolerance for simplification."),
+           &m_opts.tolerance)
+        .SetRequired()
+        .SetMinValueIncluded(0);
+    AddArg("preserve-boundary", 0,
+           _("Whether the exterior boundary should be preserved."),
+           &m_opts.preserveBoundary);
+}
+
+#if defined HAVE_GEOS &&                                                       \
+    (GEOS_VERSION_MAJOR > 3 ||                                                 \
+     (GEOS_VERSION_MAJOR == 3 && GEOS_VERSION_MINOR >= 12))
+
+/**
+ * Dataset used to read all input features into memory and simplify their
+ * geometries.
+ */
+class GDALVectorGeomSimplifyCoverageOutputDataset final : public GDALDataset
+{
+  public:
+    GDALVectorGeomSimplifyCoverageOutputDataset();
+
+    void AddLayer(std::unique_ptr<OGRLayer> poNewLayer)
+    {
+        m_layers.push_back(std::move(poNewLayer));
+    }
+
+    int GetLayerCount() override
+    {
+        return static_cast<int>(m_layers.size());
+    }
+
+    OGRLayer *GetLayer(int idx) override
+    {
+        return m_layers[idx].get();
+    }
+
+  private:
+    std::vector<std::unique_ptr<OGRLayer>> m_layers{};
+};
+
+class GDALVectorGeomSimplifyCoverageAlgorithmLayer final : public OGRLayer
+{
+  public:
+    explicit GDALVectorGeomSimplifyCoverageAlgorithmLayer(
+        OGRLayer &src,
+        const GDALVectorGeomSimplifyCoverageAlgorithm::Options &opts)
+        : m_srcLayer(src), m_opts(opts)
+    {
+    }
+
+    ~GDALVectorGeomSimplifyCoverageAlgorithmLayer() override
+    {
+        if (m_GEOSContext != nullptr)
+        {
+            for (size_t i = 0; i < m_features.size(); i++)
+            {
+                GEOSGeom_destroy_r(m_GEOSContext, m_geos_result[i]);
+            }
+            GEOSFree_r(m_GEOSContext, m_geos_result);
+            finishGEOS_r(m_GEOSContext);
+        }
+    }
+
+    int TestCapability(const char *pszCap) override
+    {
+        if (EQUAL(pszCap, OLCRandomRead))
+        {
+            return true;
+        }
+        return m_srcLayer.TestCapability(pszCap);
+    }
+
+    OGRErr IGetExtent(int iGeomField, OGREnvelope *psExtent,
+                      bool bForce) override
+    {
+        return m_srcLayer.GetExtent(iGeomField, psExtent, bForce);
+    }
+
+    OGRFeatureDefn *GetLayerDefn() override
+    {
+        return m_srcLayer.GetLayerDefn();
+    }
+
+    void ResetReading() override
+    {
+        if (!m_materialized)
+        {
+            m_srcLayer.ResetReading();
+            m_features.clear();
+        }
+        m_pos = 0;
+    }
+
+    OGRFeature *GetNextFeature() override
+    {
+        if (!m_materialized)
+        {
+            if (!Materialize())
+            {
+                return nullptr;
+            }
+        }
+
+        if (m_pos >= m_features.size())
+        {
+            return nullptr;
+        }
+
+        auto &ret = m_features[m_pos];
+        m_pos++;
+
+        return ret.release();
```

Comment:
I suspect that given you relese m_features[m_pos], if you do ResetReading() and iterate again, you'll get a null feature. Perhaps you should return ret->Clone() instead ?

To check your implementation of OGRLayer is conformant, you could run test_ogrsf on a .gdalg file. See test_gdalalg_vector_clip_test_ogrsf() for example